File: services/email-query-processing/src/types/email.ts
/**
 * Email Service Types
 * 
 * Provider-agnostic types for email operations.
 */

/**
 * Email attachment
 */
export interface EmailAttachment {
  id: string; // Provider-specific attachment ID
  filename: string;
  contentType: string;
  size: number;
  data?: Buffer; // Optional: inline data
}

/**
 * Email sender information
 */
export interface EmailSender {
  email: string;
  name?: string;
}

/**
 * Email recipients
 */
export interface EmailRecipients {
  to: string[];
  cc?: string[];
  bcc?: string[];
}

/**
 * Provider-agnostic Email model
 */
export interface Email {
  id: string; // Provider-specific ID (Gmail ID, IMAP UID, etc.)
  providerType: string; // "gmail", "imap", "sendgrid", etc.
  threadId?: string; // Thread/conversation ID (if supported)
  sender: EmailSender;
  recipients: EmailRecipients;
  subject: string;
  bodyText: string;
  bodyHtml?: string;
  receivedAt: Date;
  attachments?: EmailAttachment[];
  metadata?: Record<string, any>; // Provider-specific metadata
}

/**
 * Inbox information
 */
export interface Inbox {
  id: string;
  name: string;
  address: string;
}

/**
 * Rate limit information
 */
export interface RateLimitInfo {
  requestsPerSecond?: number;
  requestsPerMinute?: number;
  requestsPerHour?: number;
  requestsPerDay?: number;
  quotaUnits?: number; // For Gmail API quota units
}

/**
 * Provider capabilities
 */
export interface ProviderCapabilities {
  supportsWebhooks: boolean; // Can receive push notifications
  supportsPolling: boolean; // Can poll for emails
  supportsLabels: boolean; // Can use labels/folders
  supportsThreading: boolean; // Supports email threading
  maxAttachmentSize: number; // Max attachment size in bytes
  rateLimits: RateLimitInfo; // Rate limit information
  authenticationType: string; // "oauth2", "api_key", "password", etc.
}

/**
 * Watch handle for inbox watching
 */
export interface WatchHandle {
  id: string;
  stop: () => Promise<void>;
}

File: services/email-ingestion/src/types/email.py
"""
Email Service Types

Provider-agnostic types for email operations.
"""

from typing import List, Optional, Dict, Any
from datetime import datetime
from dataclasses import dataclass


@dataclass
class EmailAttachment:
    """Email attachment."""
    id: str  # Provider-specific attachment ID
    filename: str
    contentType: str
    size: int
    data: Optional[bytes] = None  # Optional: inline data


@dataclass
class EmailSender:
    """Email sender information."""
    email: str
    name: Optional[str] = None


@dataclass
class EmailRecipients:
    """Email recipients."""
    to: List[str]
    cc: Optional[List[str]] = None
    bcc: Optional[List[str]] = None


@dataclass
class Email:
    """Provider-agnostic Email model."""
    id: str  # Provider-specific ID (Gmail ID, IMAP UID, etc.)
    providerType: str  # "gmail", "imap", "sendgrid", etc.
    threadId: Optional[str] = None  # Thread/conversation ID (if supported)
    sender: EmailSender = None
    recipients: EmailRecipients = None
    subject: str = ''
    bodyText: str = ''
    bodyHtml: Optional[str] = None
    receivedAt: datetime = None
    attachments: Optional[List[EmailAttachment]] = None
    metadata: Optional[Dict[str, Any]] = None  # Provider-specific metadata


@dataclass
class Inbox:
    """Inbox information."""
    id: str
    name: str
    address: str


@dataclass
class RateLimitInfo:
    """Rate limit information."""
    requestsPerSecond: Optional[int] = None
    requestsPerMinute: Optional[int] = None
    requestsPerHour: Optional[int] = None
    requestsPerDay: Optional[int] = None
    quotaUnits: Optional[int] = None  # For Gmail API quota units


@dataclass
class ProviderCapabilities:
    """Provider capabilities."""
    supportsWebhooks: bool  # Can receive push notifications
    supportsPolling: bool  # Can poll for emails
    supportsLabels: bool  # Can use labels/folders
    supportsThreading: bool  # Supports email threading
    maxAttachmentSize: int  # Max attachment size in bytes
    rateLimits: RateLimitInfo  # Rate limit information
    authenticationType: str  # "oauth2", "api_key", "password", etc.


@dataclass
class WatchHandle:
    """Watch handle for inbox watching."""
    id: str
    
    def stop(self) -> None:
        """Stop watching inbox."""
        pass
